packages/nitro-protocol/contracts/examples/TwoOfThree.sol

```diff
@@ -0,0 +1,119 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.7.4;
+pragma experimental ABIEncoderV2;
+
+import '../interfaces/IForceMoveApp.sol';
+import '../Outcome.sol';
+import '../interfaces/IForceMove.sol';
+
+/**
+ * @dev The HashLockedSwap contract complies with the ForceMoveApp interface and implements a HashLockedSwaped payment
+ */
+contract TwoOfThree is IForceMoveApp2 {
+    struct AppData {
+        IForceMoveApp2 twoPartyApp;
+        bytes twoPartyAppData;
+    }
+
+    function validTransition(
+        VariablePart memory a, // TODO consider less confusing variable names
+        VariablePart memory b,
+        uint48 turnNumB,
+        uint256,
+        uint8 signedBy // Who has signed state b?
+    ) public override pure returns (bool) {
+        // If the new state is signed by all three participants, it is automatically considered a valid transition
+        if (signedBy == 7) return true;
+        if (signedBy == 0) return false;
+        if (signedBy < 4) {
+            // participant 2 has not signed
+            AppData memory aAppData = abi.decode(a.appData, (AppData));
+            AppData memory bAppData = abi.decode(b.appData, (AppData));
+
+            // Decode variables.
+            // Assumptions:
+            //  - single asset in this channel
+            //  - three parties in this channel
+            (
+                Outcome.AllocationItem[] memory allocationA,
+                address aAssetHolderAddress
+            ) = decode3PartyAllocation(a.outcome);
+            (
+                Outcome.AllocationItem[] memory allocationB,
+                address bAssetHolderAddress
+            ) = decode3PartyAllocation(b.outcome);
+
+            // slots for each participant unchanged
+            require(
+                allocationA[0].destination == allocationB[0].destination &&
+                    allocationA[1].destination == allocationB[1].destination &&
+                    allocationA[2].destination == allocationB[2].destination,
+                'destinations may not change'
+            );
+
+            // participant 2's slot may not change
+            require(allocationA[2].amount == allocationB[2].amount, 'p2.amt: no sig');
+
+            // slice off the two party outcome and data
+            VariablePart memory a2 = VariablePart(
+                encode2PartyAllocation(allocationA[0], allocationA[1], aAssetHolderAddress),
+                aAppData.twoPartyAppData
+            );
+            VariablePart memory b2 = VariablePart(
+                encode2PartyAllocation(allocationB[0], allocationB[1], bAssetHolderAddress),
+                bAppData.twoPartyAppData
+            );
+
+            return bAppData.twoPartyApp.validTransition(a2, b2, turnNumB, 2, signedBy);
```

Comment:
This validates the updated app data. How does one force an update in the _outcome_ of `J`, based on some update in the embedded `twoPartyApp`?

Reply:
I'm a bit confused here, but this implementation seems to couple app updates & joint channel updates. Why is `turnNumB` passed to the `twoPartyApp`'s valid transition function? How do Alice & Bob update the app channel with larger and larger turn numbers, without signing updates in the joint channel?

Reply:
The idea was to have no distinction between the app channel and the joint channel: so that Alice and Bob are always signing updates to the joint channel.

This is a bad design, however, as it doesn't allow Irene to force the channel to conclude in all cases.

Reply:
Ahhhh. That makes sense!